README.md
mailto2
=======

mailto2 is a CGI WWW form to mail gateway written in C.  It protects against spam by allowing outgoing mail to only those addresses specified in the configuration file.  It can be used to process contact forms on otherwise static sites.

mailto2 is a fork of the Debian GNU/Linux revised version of Andreas Ley's mailto program.

Installation
------------

Run `make` to build.

Run `make install` (as root) to install to `/usr/lib/cgi-bin`.

Usage
-----

Add the email addresses that should receive mail from the WWW mail gateway to `/etc/mailto.conf`, one per line. Only mail to addresses listed in `/etc/mailto.conf` will be delivered.

Create your an HTML form with `action="/cgi-bin/mailto/<address>"` and `method="post"`.  `<address>` is the recipient's email address.  You can instead provide this parameter with a (hidden) `input` tag with `name="To"`, which will have precedence over the URL-supplied parameter.  Provide a subject with an `input` tag with `name="Subject"`.

When the form is submitted, the recipient address(es) (including Cc & Bcc recipients) will be checked against the configuration file, the contents will be converted into a mail message and sent.  Other than the special fields listed below, input fields will be listed in the message body as `<name>: <content><LF>` for single line contents and as `<name>:<LF><content><LF><content><LF>.<LF>` for multiline contents.  Leading periods will be preceded by a second period (dot-stuffing).  If a field with `name="Acknowledge"` was provided, the user will be forwarded to the specified URL upon success as a confirmation to the sender.

### Special fields

The following fields have a special meaning when they are part of the form data.  The field names are case sensitive.  They won't end up in the mail body but will be used for constructing and delivering the resulting mail.

* **From** - Submitter address, will be used as `From:` header **(required)**
* **To** - Primary recipient, can also be submitted in URL as `.../<address>`
* **Cc** - Carbon copy recipient (included in mail header)
* **Bcc** - Blind carbon copy recipient (not included in mail header)
* **Subject** - Mail subject **(required)**
* **Acknowledge** - URL for redirection on success

### Example form

```html
<form action="/cgi-bin/mailto/you@example.com" method="POST">
  <p>From: <input type="text" name="From" /></p>
  <p>Subject: <input type="text" name="Cc" /></p>
  <p>Message: <input type="text" type="Message" /></p>
  <input type="hidden" name="Acknowledge" value="/special/mailto-answer.html" />
  <button type="submit">Send</button>
</form>
```

Contributing
------------

All contributions (e.g. issue reports, pull requests, questions and comments) are encouraged.

Authors
-------

* **Andreas Ley** (ley@rz.uni-karlsruhe.de) - initial author
* **Martin Schulze** (joey@infodrom.org) - Debian package
* **Evan Thompson** ([alvanson](https://github.com/alvanson)) - current maintainer

This is Differentiated Analytics Inc.'s fork of the Debian GNU/Linux prepackaged version of Andreas Ley's mailto program.

The Debian package was put together by Martin Schulze (joey@debian.org), from sources obtained from: ftp://ftp.rz.uni-karlsruhe.de/pub/net/www/tools/cgi-src/mailto.tar.gz

One mirror still seems to exist: http://www-ftp.lip6.fr/ftp/pub/www/more-tools/cgi-src/mailto.tar.gz

Both sites only contained mailto up to version 1.2.6.

The upstream packages for 1.2.7, 1.3, 1.3.1 and 1.3.2 were sent to Joey via private mail and not released to the abovely mentioned Internet address.

Version 1.3.2 is also the final upstream version which only exists on a tape library archive and isn't in use anymore.

License
-------

* Copyright (c) 1994-1999 by Andreas Ley (ley@rz.uni-karlsruhe.de)
* Copyright (c) 1998,2004 by Martin Schulze (joey@infodrom.org)
* Copyright (c) 2017 by Differentiated Analytics Inc. (info@differentiatedanalytics.ca)

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

Acknowledgements
----------------

Thank you to the following for bug reports, ideas and fixes for the original mailto project and Debian package:

* Adriano R. (anr@ime.usp.br)
* Michael Pall (pall@rz.uni-karlsruhe.de)
* Otis Gospodnetic (otisg@cobalt.middlebury.edu)
* Martin Schulze (joey@infodrom.org)
* Ryan Hayle (ryan@savvynet.com)


CHANGELOG.md
Changelog
=========

2.0.0 (2017-09-04)
------------------
* mailto now verifies all outgoing addresses against /etc/mailto.conf.
* mailto no longer supports extra data (realnames) in /etc/mailto.conf.
* Removed template mail form; only POST requests are accepted.
* Removed Copy-Self field (potential spam vector).
* Removed deprecated Reply-To field.
* Removed German translation.
* Removed Debian packaging files.
* Removed lwp-test program.
* Code and documentation updated and reformatted throughout.
* Addressed potential buffer overflow vulnerabilities.

1.3.2-3 (2010-01-16)
--------------------
* Disable included version of getline() since it is unused
* Don't use '0' flag used with '%p' gnu_printf format

1.3.2-2 (2007-03-17)
--------------------
* Extract the local mail hostname from /etc/mailname
* Fixed two compiler warnings

1.3.2-1 (2004-12-12)
--------------------
* Added support for special mail headers required when delivering mail with non-ascii character encoding

1.3.2 (1999-03-02)
------------------
* Conditional error-to.

1.3.1-1 (2004-12-11)
--------------------
* Adjusted native report on Cc
* Added section about special variables for better documentation
* Added missing \r for the smtp protocol

1.3.1 (1998-09-02)
------------------
* Fixed carbon copy option.

1.3-1 (2004-12-09)
------------------
* New license: GNU GPLv2
* Added documentation for carbon copy (Cc)
* Better note webmaster@ as address to send errors to

1.3 (1996-08-13)
----------------
* Option to send a carbon copy. (Not yet working)

1.2.7-1 (2004-12-08)
--------------------
* Fixed format string vulnerability

1.2.7 (1996-04-12)
------------------
* Added random number (pid) code to subject parsing

1.2.6-5 (2004-12-04)
--------------------
* Cleaned up some code in mailto.c
* Wrote and added a small test program
* Added an UTF-8 detection rules so that valid UTF-8 characters won't be eaten anymore

1.2.6-4 (2001-08-10)
--------------------
* Finally fixed the redirection problem that mailto had.
* Added bilingual message for HTTP relocation
* Send CRLF in the SMTP dialog instead of just LF
* Corrected some German messages
* Added information about potential mail lossage when the pipe to sendmail is closed which caused an error

1.2.6-3 (2000-01-04)
--------------------
* Don't use strip for non-strippable files

1.2.6-2 (1999-10-26)
--------------------
* Removed case-sensitivity of email addresses

1.2.6-1 (1998-09-29)
--------------------
* Added readme
* Support for SENDER/MASTER@`hostname -f`
* Support for appropriate Redirect
* Added example

1.2.6 (1996-01-30)
------------------
* Fixed address checking bug with addresses without realnames

1.2.5 (1996-01-29)
------------------
* Added debugging output
* Fixed some possible references to null-pointers

1.2.4 (1996-01-25)
------------------
* Added support for machines lacking strerror(3C)

1.2.3 (1995-09-27)
------------------
* Gateway identifies with Sender: instead of From:
* User address now specified in From: field, Reply-To alias for From

1.2.2 (1995-09-05)
------------------
* Added <TAB> to the address file separators

1.2.1 (1995-08-02)
------------------
* X- postfix for Addr, Host, Ident and User headers.

1.2 (1995-06-01)
----------------
* Option to send a copy of the generated mail to oneself.
* Fixed bug when client sends <CR><LF> instead of <LF> only.

1.1.1 (1995-01-23)
------------------
* Option to show generated mail in automatic acknowledgement page.

1.1 (1995-01-19)
----------------
* Added evaluation of INPUT fields for To, Subject and Acknowledge.
* Default FORM now generates INPUT fields instead of URL parameters.

1.0.2 (1994-08-23)
------------------
* Fixed location bug.
* Supplies default FORM on method GET now.

1.0.1 (1994-06-10)
------------------
* Fixed documentation bug.

1.0 (1994-06-10)
----------------
* Initial version


util.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define LF 10
#define CR 13

#define HAS_GETLINE  /* Don't include included getline function */

void getword(char *word, char *line, char stop) {
    int x = 0,y;

    for(x=0;((line[x]) && (line[x] != stop));x++)
        word[x] = line[x];

    word[x] = '\0';
    if(line[x]) ++x;
    y=0;

    while((line[y++] = line[x++]));
}

char *makeword(char *line, char stop) {
    int x = 0,y;
    char *word = (char *) malloc(sizeof(char) * (strlen(line) + 1));

    for(x=0;((line[x]) && (line[x] != stop));x++)
        word[x] = line[x];

    word[x] = '\0';
    if(line[x]) ++x;
    y=0;

    while((line[y++] = line[x++]));
    return word;
}

char *fmakeword(FILE *f, char stop, int *cl) {
    int wsize;
    char *word;
    int ll;

    wsize = 102400;
    ll=0;
    word = (char *) malloc(sizeof(char) * (wsize + 1));

    while(1) {
        word[ll] = (char)fgetc(f);
        if(ll==wsize) {
            word[ll+1] = '\0';
            wsize+=102400;
            word = (char *)realloc((void*)word,sizeof(char)*(wsize+1));
        }
        --(*cl);
        if((word[ll] == stop) || (feof(f)) || (!(*cl))) {
            if(word[ll] != stop) ll++;
            word[ll] = '\0';
            return word;
        }
        ++ll;
    }
}

char x2c(char *what) {
    register char digit;

    digit = (what[0] >= 'A' ? ((what[0] & 0xdf) - 'A')+10 : (what[0] - '0'));
    digit *= 16;
    digit += (what[1] >= 'A' ? ((what[1] & 0xdf) - 'A')+10 : (what[1] - '0'));
    return(digit);
}

void unescape_url(char *url) {
    register int x,y;

    for(x=0,y=0;url[y];++x,++y) {
        if((url[x] = url[y]) == '%') {
            url[x] = x2c(&url[y+1]);
            y+=2;
        }
    }
    url[x] = '\0';
}

void plustospace(char *str) {
    register int x;

    for(x=0;str[x];x++) if(str[x] == '+') str[x] = ' ';
}

int rind(char *s, char c) {
    register int x;
    for(x=strlen(s) - 1;x != -1; x--)
        if(s[x] == c) return x;
    return -1;
}

#ifndef HAS_GETLINE
int getline(char *s, int n, FILE *f) {
    register int i=0;

    while(1) {
        s[i] = (char)fgetc(f);

        if(s[i] == CR)
            s[i] = fgetc(f);

        if((s[i] == 0x4) || (s[i] == LF) || (i == (n-1))) {
            s[i] = '\0';
            return (feof(f) ? 1 : 0);
        }
        ++i;
    }
}
#endif

void send_fd(FILE *f, FILE *fd)
{
    char c;

    while (1) {
        c = fgetc(f);
        if(feof(f))
            return;
        (void)fputc(c,fd);
    }
}


mailto_german.h
/* General page layout definitions */

#define	TITLE	"WWW Mail Gateway"

/* Everything's fine */

#define	SUCCESS_HEADER	NULL
#define	SUCCESS_SELF	"Ihre Meldung wurde an %s und eine Kopie an %s abgeschickt."
#define	SUCCESS_ALSO	", %s"
#define	SUCCESS_COPY	" und eine Kopie an %s abgeschickt.\n"
#define	SUCCESS_FROM	"<P><CODE>From:</CODE> %s\n"
#define	SUCCESS_SUBJECT	"<BR><CODE>Subject:</CODE> %s\n"
#define	SUCCESS_SENDER	"<BR><CODE>Sender:</CODE> %s@%s\n"
#define	SUCCESS_TO	"<BR><CODE>To:</CODE> %s\n"
#define	SUCCESS_CC	"<BR><CODE>Cc:</CODE> %s\n"
#define	SUCCESS_BCC	"<BR><CODE>Bcc:</CODE> %s\n"
#define	SUCCESS_REPLYTO	"<BR><CODE>Reply-To:</CODE> %s\n"
#define	SUCCESS_ADDR	"<BR><CODE>X-Addr:</CODE> %s\n"
#define	SUCCESS_HOST	"<BR><CODE>X-Host:</CODE> %s\n"
#define	SUCCESS_IDENT	"<BR><CODE>X-Ident:</CODE> %s\n"
#define	SUCCESS_USER	"<BR><CODE>X-User:</CODE> %s\n"
#define	SUCCESS_SINGLE	"<P><CODE>%s:</CODE> %s\n"
#define	SUCCESS_MULTI	"<P><CODE>%s:</CODE>\n"
#define	SUCCESS_LINE	"<BR>%s\n"

/* Error messages */

#define	ERROR	"Fehler"
#define	ERROR_FATAL	"<P>Ein solcher Fehler darf eigentlich nicht auftreten. Bitte setzen Sie sich mit <EM>%s@%s</EM> in Verbindung und erl&auml;utern Sie, wie es zu diesem Fehler gekommen ist. Interessant ist insbesondere, von welcher Seite Sie diese Fehlermeldung erhalten, was Sie eingegeben haben und welche Software Sie verwenden."
#define	ERROR_REQUEST_METHOD	"Es wurde ein falscher REQUEST_METHOD-Aufruf (%s) verwendet, erwartet wurde %s."
#define	ERROR_CONTENT_TYPE	"Es wurde ein falscher CONTENT_TYPE (%s) &uuml;bergeben, erwartet wurde %s."
#define	ERROR_CONTENT_LENGTH	"CONTENT_LENGTH ist null."
#define	ERROR_ADDRESS	"Es fehlen die Angaben &uuml;ber Mail-Adresse und Subject."
#define	ERROR_READ	"Fehler beim Lesen der Adressdatei."
#define	ERROR_POPEN	"Fehler beim Versenden der Meldung:<P><CODE>popen(\"%s\"): %s</CODE>"
#define	ERROR_PCLOSE	"Fehler beim Versenden der Meldung:<P><CODE>%s</CODE> liefert exit code %d.  Mail kann verloren gegangen sein."
#define	BAD_ADDRESS	"Die angegebene Adresse %s ist nicht in der Liste der erlaubten Adressen eingetragen. Um die Benutzer vor anonymen Mails zu sch&uuml;tzen, m&uuml;ssen potentielle Empf&auml;nger sich (via Email an %s@%s) in diese Liste eintragen lassen."


mailto.c
/*
** mailto2 - WWW form to mail gateway
**
** Copyright (c) 1994-1999 by Andreas Ley <ley@rz.uni-karlsruhe.de>
** Copyright (c) 1998,2004 by Martin Schulze <joey@infodrom.org>
** Copyright (c) 2017 by Differentiated Analytics Inc.
**                       <info@differentiatedanalytics.ca>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License along
** with this program; if not, write to the Free Software Foundation, Inc.,
** 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "mailto_english.h"

#include <netdb.h>
#include <unistd.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* File locations */

#define ADDRESSES "/etc/mailto.conf"
#define MAILCMD "/usr/lib/sendmail -bs >/dev/null"

/* General definitions */

#define POSTMASTER "postmaster"
#define WEBMASTER "webmaster"
#define SENDER "www"

/* Global variables */

#define BUFSIZE 8192
#define MAXFIELDS 1024

char header[] = "mailto2 2.0.0\n"
"(c) 1994-1999 by Andreas Ley <Andreas.Ley@rz.uni-karlsruhe.de>\n"
"(c) 1998,2004 by Martin Schulze <joey@infodrom.org>\n"
"(c) 2017 by Differentiated Analytics Inc. <info@differentiatedanalytics.ca>\n";

char fqdn[256];
char mailname[256];
char *htag[MAXFIELDS], *hval[MAXFIELDS];
size_t max;
int debug = 0;


/* print header of an HTML response */
void show_header(char *subtitle, char *description)
{
    if (debug) fprintf(stderr, "show_header(\"%s\",\"%s\")\n", subtitle, description);

    printf("Content-type: text/html\n\n");
    printf("<!DOCTYPE html>\n");
    printf("<!--\n%s-->\n", header);
    printf("<html>\n");
    printf("  <head>\n");
    printf("    <meta charset=\"utf-8\" />\n");
    printf("    <title>%s%s%s</title>\n", TITLE, subtitle ? " - " : "" , subtitle ? subtitle : "");
    printf("  </head>\n");
    printf("  <body>\n");
    printf("    <h1>%s</h1>\n", subtitle ? subtitle : TITLE);
    if (description) {
        printf("<p>");
        printf(description, WEBMASTER, mailname);
        printf("</p>\n");
    }
}


/* print footer (trailer) of an HTML response */
void show_trailer()
{
    if (debug) fprintf(stderr, "show_trailer()\n");

    printf("  </body>\n");
    printf("</html>\n");
}


/* print an error page (wrong user input) */
void show_error(char *error)
{
    if (debug) fprintf(stderr, "show_error(\"%s\")\n", error);

    show_header(ERROR, error);
    show_trailer();
    exit(0);
}


/* print an error page (internal malfunction) */
void show_fatal(char *error)
{
    if (debug) fprintf(stderr, "show_fatal(\"%s\")\n", error);

    show_header(ERROR, error);
    printf("<p>");
    printf(ERROR_FATAL, WEBMASTER, mailname);
    printf("</p>\n");
    show_trailer();
    exit(0);
}


/* redirect to supplied location */
void show_location(char *location)
{
    if (debug) fprintf(stderr, "show_location(\"%s\")\n", location);

    printf("Status: 302 Found\n");
    printf("Location: %s\n\n", location);
}


/* strip control and invalid (non utf-8) characters */
char *secure(char *text)
{
    char *src, *dest;
    size_t i;

    if (debug) fprintf(stderr, "secure(\"%s\")", text);

    if (text) {
        for (src=text, dest=text; *src; /* nop */) {
            if (*src >= 0x20 && *src < 0x7f) { /* 1 byte, skip control characters and <DEL> (0x7f) */
                *dest++ = *src++;
            } else if ((*src & 0b11100000) == 0b11000000 &&     /* 2 byte prefix */
                        (*(src+1) & 0b11000000) == 0b10000000) {    /* 2nd byte valid */
                for (i=0; i<2; i++) *dest++ = *src++;
            } else if ((*src & 0b11110000) == 0b11100000 &&     /* 3 byte prefix */
                        (*(src+1) & 0b11000000) == 0b10000000 &&    /* 2nd byte valid */
                        (*(src+2) & 0b11000000) == 0b10000000) {    /* 3rd byte valid */
                for (i=0; i<3; i++) *dest++ = *src++;
            } else if ((*src & 0b11111000) == 0b11110000 &&     /* 4 byte prefix */
                        (*(src+1) & 0b11000000) == 0b10000000 &&    /* 2nd byte valid */
                        (*(src+2) & 0b11000000) == 0b10000000 &&    /* 3rd byte valid */
                        (*(src+3) & 0b11000000) == 0b10000000) {    /* 4th byte valid */
                for (i=0; i<4; i++) *dest++ = *src++;
            } else {
                src++;
            }
        }
        *dest = '\0';
    }

    if (debug) fprintf(stderr, "=\"%s\"\n", text);

    return text;
}


/* replace <CR><LF> with <LF> only */
char *stripcr(char *text)
{
    char *src, *dest;

    if (debug) fprintf(stderr, "stripcr(\"%s\")\n", text);

    if (text) {
        for (src=text, dest=text; *src; src++) {
            if (src[0] != '\r' || src[1] != '\n') {
                *dest++ = *src;
            }
        }
        *dest = '\0';
    }
    return text;
}


/* check for valid address */
char *checkaccess(char *address)
{
    FILE *f;
    char line[1024];
    char *ptr;

    if (debug) fprintf(stderr, "checkaccess(\"%s\")\n", address);

    if ((f = fopen(ADDRESSES, "r"))) {
        while (fgets(line, sizeof(line), f)) {
            if (*line != '#') {    /* skip comments */
                /* strip trailing newline */
                if ((ptr = strchr(line, '\n'))) {
                    *ptr = '\0';
                }
                if (debug) fprintf(stderr, "checkaccess: \"%s\"\n", line);
                if (strcmp(address, line) == 0) {
                    /* address in file */
                    fclose(f);
                    if (debug) fprintf(stderr, "checkaccess: valid \"%s\"\n", line);
                    return address;
                }
            }
        }
        /* address not in file */
        fclose(f);
        snprintf(line, sizeof(line), BAD_ADDRESS, address, POSTMASTER, mailname);
        show_error(line);
        return NULL;
    } else {
        show_fatal(ERROR_READ);
        return NULL;
    }
}


/* actually send mail to recipient(s) */
void mailto(char *address, char *cc, char *bcc, char *subject, char *from)
{
    FILE *p;
    int i;
    char *ptr, *nptr;
    int retval;
    char error[BUFSIZE];

    if (debug) fprintf(stderr, "mailto(\"%s\",\"%s\",\"%s\",\"%s\",\"%s\")\n", address, cc, bcc, subject, from);

    if ((p = popen(MAILCMD, "w"))) {
        fprintf(p, "HELO %s\r\n", fqdn);
        fprintf(p, "MAIL FROM:<%s@%s>\r\n", SENDER, mailname);
        fprintf(p, "RCPT TO:<%s>\r\n", address);
        if (cc) {
            fprintf(p, "RCPT TO:<%s>\r\n", cc);
        }
        if (bcc) {
            fprintf(p, "RCPT TO:<%s>\r\n", bcc);
        }
        fprintf(p, "DATA\r\n");
        /* message headers */
        fprintf(p, "From: %s\r\n", from);
        fprintf(p, "Sender: %s@%s\r\n", SENDER, mailname);
        fprintf(p, "To: %s\r\n", address);
        if (cc) {
            fprintf(p, "Cc: %s\r\n", cc);
        }
        fprintf(p, "Subject: %s\r\n", subject);
        fprintf(p, "Content-Type: text/plain; charset=utf-8\r\n");
        fprintf(p, "Content-Disposition: inline\r\n");
        fprintf(p, "Content-Transfer-Encoding: 8bit\r\n");
        if ((ptr = getenv("REMOTE_ADDR")) && *ptr) {
            fprintf(p, "X-Addr: %s\r\n", ptr);
        }
        if ((ptr = getenv("REMOTE_HOST")) && *ptr) {
            fprintf(p, "X-Host: %s\r\n", ptr);
        }
        if ((ptr = getenv("REMOTE_IDENT")) && *ptr) {
            fprintf(p, "X-Ident: %s\r\n", ptr);
        }
        if ((ptr = getenv("REMOTE_USER")) && *ptr) {
            fprintf(p, "X-User: %s\r\n", ptr);
        }
        fprintf(p, "\r\n");
        /* message body: print all the fields preceeded by their name */
        if (max > MAXFIELDS) {
            max = MAXFIELDS;
        }
        for (i = 0; i < max; i++) {
            /* tag dot-stuffing */
            if (*htag[i] == '.') {
                fprintf(p, ".");
            }
            if ((nptr = strchr(hval[i], '\n'))) {
                /* multiline format */
                fprintf(p, "%s:\r\n", htag[i]);
                ptr = hval[i];
                do {
                    /* terminate line */
                    *nptr = '\0';
                    /* line dot-stuffing */
                    if (*ptr == '.') {
                        fprintf(p, ".");
                    }
                    fprintf(p, "%s\r\n", secure(ptr));
                    /* restore newline */
                    *nptr = '\n';
                    /* get next line */
                    ptr = nptr + 1;
                } while ((nptr = strchr(ptr, '\n')));
                /* final line */
                if (*ptr == '.') {
                    fprintf(p, ".");
                }
                fprintf(p, "%s\r\n", secure(ptr));
                fprintf(p, "..\r\n");
            } else {
                /* single line format */
                fprintf(p, "%s: %s\r\n", htag[i], secure(hval[i]));
            }
        }
        /* end message */
        fprintf(p, ".\r\n");
        fprintf(p, "QUIT\r\n");
        /* check return value for success */
        retval = pclose(p);
        if (retval) {
            snprintf(error, sizeof(error), ERROR_PCLOSE, MAILCMD, retval);
            show_fatal(error);
        }
    } else {
        snprintf(error, sizeof(error), ERROR_POPEN, MAILCMD, strerror(errno));
        show_fatal(error);
    }
}


void usage(char *image)
{
    fprintf(stderr, "Usage: %s [-h] [-v]\n", image);
    exit(1);
}


/* get the fqdn of this machine */
void getfqdn(char *name, size_t len)
{
    struct addrinfo hints;
    struct addrinfo *result;

    gethostname(name, len-1);
    /* \0 not guaranteed by gethostname() if len exceeded */
    name[len-1] = '\0';
    /* get canonical name */
    memset(&hints, 0, sizeof(struct addrinfo));
    hints.ai_family = AF_UNSPEC;    /* allow IPv4 or IPv6 */
    hints.ai_flags = AI_CANONNAME;  /* get canonical name */
    if (getaddrinfo(name, NULL, &hints, &result) == 0) {
        /* canonical name returned in first addrinfo structure */
        if (result != NULL) {
            snprintf(name, len, "%s", result->ai_canonname);
        }
    }
    freeaddrinfo(result);
}


/* get the local mailname for later insertion */
void getmailname(char *name, size_t len, char *fallback)
{
    FILE *f;
    size_t sl;

    name[0] = '\0';

    /* try /etc/mailname first (Debian specific) */
    if ((f = fopen("/etc/mailname", "r")) != NULL) {
        if (fgets(name, len, f)) {
            /* strip trailing newline */
            sl = strlen(name);
            if (sl > 0 && name[sl-1] == '\n') {
                name[sl-1] = '\0';
            }
        }
        fclose(f);
    }
    /* fall back to supplied name */
    if (!strlen(name)) {
        snprintf(name, len, "%s", fallback);
    }
}


/* extract tokens from streams */
char *fntok_r(FILE *f, const char *delim, size_t *n)
{
    size_t len, i;
    char *token;
    char c;

    len = BUFSIZE, i = 0;
    token = malloc(len);

    while (*n > 0 && !feof(f)) {
        /* read next character */
        c = (char)fgetc(f);
        (*n)--;
        /* look for c in delimiters */
        if (strchr(delim, c)) {
            /* terminate and return token */
            token[i] = '\0';
            return token;
        } else {
            /* append character to current token */
            token[i] = c;
        }
        i++;
        /* resize token buffer if necessary */
        if (i >= len-1) {
            len += BUFSIZE;
            token[i] = '\0';
            token = realloc(token, len);
        }
    }

    token[i] = '\0';
    /* if no token remains, return NULL */
    if (*token == '\0') {
        free(token);
        token = NULL;
    }
    return token;
}


char x2c(char *x) {
    char c;

    c = (x[0] >= 'A' ? ((x[0] & 0xdf) - 'A') + 10 : (x[0] - '0'));
    c *= 16;
    c += (x[1] >= 'A' ? ((x[1] & 0xdf) - 'A') + 10 : (x[1] - '0'));

    return c;
}


void urldecode(char *url) {
    size_t i;   /* current index of input */
    size_t j;   /* current index of output, j <= i */

    for (i=0, j=0; url[i]; i++, j++) {
        if (url[i] == '+') {
            url[j] = ' ';
        } else if (url[i] == '%') {
            url[j] = x2c(&url[i+1]);
            i += 2;
        } else {
            url[j] = url[i];
        }
    }
    url[j] = '\0';
}


int main(int argc, char *argv[])
{
    int c;
    char *address, *from, *cc, *bcc, *subject, *location;
    char *ptr, *nptr;
    char *tag, *val;
    size_t cl, i;
    char *saveptr;

    while ((c = getopt(argc, argv, "Dvh?")) != EOF) {
        switch ((char)c) {
            case 'D':
                debug++;
                break;
            case 'v':
                fprintf(stderr, header);
                exit(0);
                break;
            case 'h':
                fprintf(stderr, header);
                /* fall through */
            case '?':
                usage(argv[0]);
                break;
        }
    }

    address = NULL;
    subject = NULL;
    location = NULL;
    from = SENDER;
    cc = NULL;
    bcc = NULL;

    getfqdn(fqdn, sizeof(fqdn));
    if (debug) fprintf(stderr, "fqdn=\"%s\"\n", fqdn);

    getmailname(mailname, sizeof(mailname), fqdn);
    if (debug) fprintf(stderr, "mailname=\"%s\"\n", mailname);

    /* extract recipient address from URL */
    ptr = getenv("PATH_INFO");
    if (debug) fprintf(stderr, "PATH_INFO=\"%s\"\n", ptr);
    if (ptr) {
        /* skip leading slash */
        if (*ptr == '/') {
            ptr++;
        }
        address = secure(ptr);
        if (debug) fprintf(stderr, "address=\"%s\"\n", address);
    }

    /* only accept POST requests */
    ptr = getenv("REQUEST_METHOD");
    if (debug) fprintf(stderr, "REQUEST_METHOD=\"%s\"\n", ptr);
    if (!ptr || strcmp(ptr, "POST") != 0) {
        show_fatal(ERROR_REQUEST_METHOD);
    } else {
        /* parse form content, extract special keywords */
        ptr = getenv("CONTENT_LENGTH");
        cl = ptr ? atoi(ptr) : 0;
        if (cl <= 0) {
            show_fatal(ERROR_CONTENT_LENGTH);
        }
        max = 0;
        while ((val = fntok_r(stdin, "&", &cl))) {
            if (*val) {
                urldecode(val);
                tag = secure(strtok_r(val, "=", &saveptr));
                val = strtok_r(NULL, "=", &saveptr);
                if (val) {
                    if (strcmp(tag, "To") == 0) {
                        address = secure(val);
                    } else if (strcmp(tag, "Subject") == 0) {
                        subject = secure(val);
                        if (debug) fprintf(stderr, "subject=\"%s\"\n", subject);
                    } else if (strcmp(tag, "From") == 0) {
                        from = secure(val);
                    } else if (strcmp(tag, "Cc") == 0) {
                        cc = secure(val);
                    } else if (strcmp(tag, "Bcc") == 0) {
                        bcc = secure(val);
                    } else if (strcmp(tag, "Acknowledge") == 0) {
                        location = secure(val);
                    } else {
                        /* treat htag, hval as circular buffers */
                        htag[max % MAXFIELDS] = tag;
                        hval[max % MAXFIELDS] = stripcr(val);
                        max++;
                    }
                }
            }
        }

        /* ensure address and subject supplied */
        if (!address || !*address || !subject || !*subject) {
            show_error(ERROR_ADDRESS);
        }

        /* check for valid addresses */
        address = checkaccess(address);
        cc = (cc && *cc) ? checkaccess(cc) : NULL;
        bcc = (bcc && *bcc) ? checkaccess(bcc) : NULL;

        /* send mail */
        mailto(address, cc, bcc, subject, from);

        /* redirect user to acknowledgement page */
        if (location && *location) {
            show_location(location);
        } else {
            /* generate an acknowledgement page */
            show_header(SUCCESS_HEADER, NULL);
            printf("<p>");
            printf(SUCCESS_SENT, address);
            if (cc && bcc) {
                printf(SUCCESS_ALSO, cc);
                printf(SUCCESS_COPY, bcc);
            } else if (cc) {
                printf(SUCCESS_COPY, cc);
            } else if (bcc) {
                printf(SUCCESS_COPY, bcc);
            } else {
                printf(".\n");
            }
            printf("</p>\n<p>\n");
            printf(SUCCESS_FROM, from);
            printf(SUCCESS_SENDER, SENDER, mailname);
            printf(SUCCESS_TO, address);
            if (cc) {
                printf(SUCCESS_CC, cc);
            }
            if (bcc) {
                printf(SUCCESS_BCC, bcc);
            }
            printf(SUCCESS_SUBJECT, subject);
            if ((ptr = getenv("REMOTE_ADDR")) && *ptr) {
                printf(SUCCESS_ADDR, ptr);
            }
            if ((ptr = getenv("REMOTE_HOST")) && *ptr) {
                printf(SUCCESS_HOST, ptr);
            }
            if ((ptr = getenv("REMOTE_IDENT")) && *ptr) {
                printf(SUCCESS_IDENT, ptr);
            }
            if ((ptr = getenv("REMOTE_USER")) && *ptr) {
                printf(SUCCESS_USER, ptr);
            }
            printf("</p>\n");
            for (i=0; i < max; i++) {
                if ((nptr = strchr(hval[i], '\n'))) {
                    /* multiline format */
                    printf(SUCCESS_MULTI_START, htag[i]);
                    ptr = hval[i];
                    do {
                        *nptr = '\0';
                        printf(SUCCESS_MULTI_LINE, secure(ptr));
                        ptr = nptr + 1;
                    } while ((nptr = strchr(ptr, '\n')));
                    /* final line */
                    printf(SUCCESS_MULTI_END, secure(ptr));
                }
                else {
                    /* single line format */
                    printf(SUCCESS_SINGLE, htag[i], secure(hval[i]));
                }
            }
            show_trailer();
        }
    }

    return 0;
}
